src/udev/udev-node.c

```diff
@@ -246,129 +238,96 @@ static int update_timestamp(sd_device *dev, const char *path, struct stat *prev)
          * timestamp of the stack directory is always increased at least in the above step 5, so A can
          * detect the update. */
 
-        if ((prev->st_mode & S_IFMT) == 0)
-                return 0; /* Does not exist, or previous stat() failed. */
-
         for (unsigned i = 0; i < UPDATE_TIMESTAMP_MAX_RETRIES; i++) {
                 struct stat st;
 
-                if (stat(path, &st) < 0)
+                if (fstat(fd, &st) < 0)
                         return -errno;
 
                 if (!stat_inode_unmodified(prev, &st))
                         return 0;
 
                 log_device_debug(dev,
-                                 "%s is modified, but its timestamp is not changed, "
-                                 "updating timestamp after 10ms.",
-                                 path);
+                                 "Stack directory is modified, but its timestamp is not changed, "
+                                 "updating timestamp after 10ms.");
 
                 (void) usleep(10 * USEC_PER_MSEC);
-                if (utimensat(AT_FDCWD, path, NULL, 0) < 0)
+                if (futimens(fd, NULL) < 0)
                         return -errno;
         }
 
         return -ELOOP;
 }
 
-static int update_stack_directory(sd_device *dev, const char *dirname, bool add) {
-        _cleanup_free_ char *filename = NULL, *data = NULL, *buf = NULL;
-        const char *devname, *id;
-        struct stat st = {};
-        int priority, r;
+static int stack_directory_update(sd_device *dev, int fd, bool add) {
+        struct stat st;
+        const char *id;
+        int r;
 
         assert(dev);
-        assert(dirname);
+        assert(fd >= 0);
 
         r = device_get_device_id(dev, &id);
         if (r < 0)
-                return log_device_debug_errno(dev, r, "Failed to get device id: %m");
-
-        filename = path_join(dirname, id);
-        if (!filename)
-                return log_oom_debug();
+                return r;
 
-        if (!add) {
-                int unlink_error = 0, stat_error = 0;
+        if (fstat(fd, &st) < 0)
+                return -errno;
 
-                if (stat(dirname, &st) < 0) {
-                        if (errno == ENOENT)
-                                return 0; /* The stack directory is already removed. That's OK. */
-                        stat_error = -errno;
-                }
+        if (add) {
+                _cleanup_free_ char *data = NULL, *buf = NULL;
+                const char *devname;
+                int priority;
 
-                if (unlink(filename) < 0)
-                        unlink_error = -errno;
+                r = sd_device_get_devname(dev, &devname);
+                if (r < 0)
+                        return r;
 
-                if (rmdir(dirname) >= 0 || errno == ENOENT)
```

Comment:
so, hmm, if this is not cleaned up anymore, then these might just accumulate without bounds, no?

i.e. i.e. consider `/dev/disk/by-uuid/` symlinks. each time a new device pops up, we'll create one stack dir for that, and if that never gets cleaned up, then eventually /run is filled up for good? what am i missing?

Reply:
Added a cleanup logic by the last commit.